src/data/Rank.ts

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2012-2017  Online-Go.com
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// A player's rank.
+export interface Rank {
+    level: number;
+    type: "Kyu" | "Dan" | "Pro";
+}
+
+export interface AmateurRank extends Rank {
+    type: "Kyu" | "Dan";
+}
+
+export interface ProfessionalRank extends Rank {
+    type: "Pro";
+}
+
+
+
+// Run-time type checks for ranks.
+export function is_amateur(rank: Rank): rank is AmateurRank {
+    return rank.type === "Kyu" || rank.type === "Dan";
+}
+
+export function is_professional(rank: Rank): rank is ProfessionalRank {
+    return rank.type === "Pro";
+}
+
+
+
+// Utility functions to create ranks.
+export function kyu(level: number): AmateurRank {
+    return validate_rank({level: level, type: "Kyu"});
+}
+
+export function dan(level: number): AmateurRank {
+    return validate_rank({level: level, type: "Dan"});
+}
+
+export function pro(level: number): ProfessionalRank {
+    return validate_rank({level: level, type: "Pro"});
+}
+
+
+
+// Convert between ranks and strings.
+export function rank_long_string(rank: Rank): string {
+    rank = validate_rank(rank);
+    return rank.level + " " + rank.type;
+}
+
+export function rank_short_string(rank: Rank): string {
+    rank = validate_rank(rank);
+    return rank.level + rank.type[0].toLowerCase();
+}
+
+export function parse_rank(rank_string: string): Rank | void {
```

Comment:
I don't think this will work like you desire when the string is coming in as a translated string, as would currently be the case with your usage of `parse_rank` https://github.com/BarneyStratford/online-go.com/blob/types/src/components/ChallengeModal/ChallengeModal.tsx#L545 https://github.com/BarneyStratford/online-go.com/blob/types/src/components/ChallengeModal/ChallengeModal.tsx#L744-L748